File: src/data/courses/dsa/graphs/mst-advanced/kruskals-mst/code/python/kruskals_mst.py
# ============================================================
# Kruskal's MST
# ============================================================
class Solution:
    def kruskalMST(self, n, edges):
        parent = list(range(n)); rank = [0]*n
        def find(x):
            if parent[x]!=x: parent[x]=find(parent[x])
            return parent[x]
        def union(x,y):
            rx,ry=find(x),find(y)
            if rx==ry: return False
            if rank[rx]<rank[ry]: rx,ry=ry,rx
            parent[ry]=rx; if rank[rx]==rank[ry]: rank[rx]+=1
            return True
        edges.sort(key=lambda e: e[2])  # Sort by weight
        total = count = 0
        for u, v, w in edges:
            if union(u, v): total += w; count += 1
            if count == n - 1: break
        return total

File: src/data/courses/dsa/graphs/mst-advanced/kruskals-mst/code/java/KruskalsMst.java
// ============================================================
// Kruskal's MST
// ============================================================
import java.util.*;
class Solution {
    int[] pa, rk;
    int find(int x){return pa[x]==x?x:(pa[x]=find(pa[x]));}
    boolean unite(int x,int y){int a=find(x),b=find(y);if(a==b)return false;if(rk[a]<rk[b]){int t=a;a=b;b=t;}pa[b]=a;if(rk[a]==rk[b])rk[a]++;return true;}
    public int kruskalMST(int n, int[][] edges) {
        pa=new int[n]; rk=new int[n]; for(int i=0;i<n;i++) pa[i]=i;
        Arrays.sort(edges, (a,b)->a[2]-b[2]);
        int total = 0, cnt = 0;
        for (int[] e : edges)
            if (unite(e[0],e[1])) { total+=e[2]; if(++cnt==n-1) break; }
        return total;
    }
}
